### PR TITLE
8299015: Ensure that HttpResponse.BodySubscribers.ofFile writes all bytes

### PR DESCRIPTION
**Pull request by ChrisHegarty**
The implementation of `BodySubscribers.ofFile` does not correctly handle the case of a short gathering write. The write should be retried until all the bytes have been written to the underlying file.

This is generally not an issue that would occur very frequently, since the `IOV_MAX` is commonly 1024 on Mac/Linux, and 16 on Windows. However, it is clearly a bug, and should be fixed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299015](https://bugs.openjdk.org/browse/JDK-8299015): Ensure that HttpResponse.BodySubscribers.ofFile writes all bytes

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to [987aabaf](https://git.openjdk.org/jdk/pull/11722/files/987aabaffbd430aea1562b9927514b7496a75613)
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - Committer) ⚠️ Review applies to [2aea8f01](https://git.openjdk.org/jdk/pull/11722/files/2aea8f01703233ba7f711e2b483f2b6ad81cd844)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [2aea8f01](https://git.openjdk.org/jdk/pull/11722/files/2aea8f01703233ba7f711e2b483f2b6ad81cd844)

### Contributors
 * Daniel Jeliński `<djelinski@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11722/head:pull/11722` \
`$ git checkout pull/11722`

Update a local copy of the PR: \
`$ git checkout pull/11722` \
`$ git pull https://git.openjdk.org/jdk pull/11722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11722`

View PR using the GUI difftool: \
`$ git pr show -t 11722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11722.diff">https://git.openjdk.org/jdk/pull/11722.diff</a>

</details>

### CODE REVIEW

**Comment by dweiss** [comment]
LGTM (not a reviewer). Thanks for adding a proper test, @ChrisHegarty .

**Comment by dfuch** [comment]
Hey @ChrisHegarty glad to see you back :-)
The proposed changes look reasonable. I will run them in the CI and approve if the tests come back clean.

I am a little puzzled by the description of the issue however as AFAIR the SocketTube will not produce a list that contains more than three ByteBuffers. So unless I'm mistaken the 1024 limit in one onNext() call should never be reached (at least when invoked by the HttpClient stack). I agree this is a bug though, and it should be fixed.

**Comment by ChrisHegarty** [comment]
> Hey @ChrisHegarty glad to see you back :-) The proposed changes look reasonable. I will run them in the CI and approve if the tests come back clean.

Thanks @dfuch.

> I am a little puzzled by the description of the issue however as AFAIR the SocketTube will not produce a list that contains more than three ByteBuffers. So unless I'm mistaken the 1024 limit in one onNext() call should never be reached (at least when invoked by the HttpClient stack).

If the server sends the response as chunked, and say 1 byte at of response body at a time, then several of these small chunks are effectively merged (by the kernel) into a single packet. On the receiving side, the HTTP Client can receive these in say even one byte buffer that contains several hundreds or even thousands of these chunks. This gets passed to the chunked response handler that parses them and creates a List of byte buffers, one for each small chunk, so there can be many many of these that eventually get sent the response subscriber.

I managed to intermittently reproduce this on my Mac, by hacking on the test - no pretty. I guess the test could be updated to try to provoke this with the HTTP Client, but it doesn't seem worth it, if we accept that it just needs to be fixed.

**Comment by dfuch** [comment]
Ah! I see. A big byte buffer read from the socket might be split into a list of small ByteBuffer chunks at the higher level, when parsing HTTP/1.1 chunks.

> I managed to intermittently reproduce this on my Mac, by hacking on the test - not pretty. I guess the test could be updated to try to provoke this with the HTTP Client, but it doesn't seem worth it, if we accept that it just needs to be fixed.

I agree that there's no need. The test you have is enough and appropriate for the fix.

**Comment by dweiss** [comment]
This approximates and reproduces my original scenario, if you're curious,
Daniel:
https://gist.github.com/dweiss/9703cfa2713a7fc211a0018a72267d90

I get something like this in the output:
Corrupted line received: 683
Corrupted line received: 351
Corrupted line received: line: 2662
Corrupted line received: ne: 4628
Corrupted line received: 8560
Missed line range: 16-683
Missed line range: 699-930
Missed line range: 947-1351
Missed line range: 1367-1396
Missed line range: 1413-1574
... [lots more]

Dawid

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 1]
```diff
@@ -281,8 +281,13 @@ public void onSubscribe(Flow.Subscription subscription) {
 
         @Override
         public void onNext(List<ByteBuffer> items) {
+            int size = Utils.remaining(items, Integer.MAX_VALUE);
             try {
-                out.write(items.toArray(Utils.EMPTY_BB_ARRAY));
+                long bytesWritten = 0;
+                ByteBuffer[] buffers = items.toArray(Utils.EMPTY_BB_ARRAY);
+                while (bytesWritten < size) {
+                    bytesWritten += out.write(buffers);
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:289]: Should we do something special here if no bytes are written, to avoid a busy loop?

Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:289]: > Should we do something special here if no bytes are written, to avoid a busy loop?

I wonder how / if this could occur in practice? From what I see elsewhere the simple, "don't check no bytes written", loop is common.  ( It shouldn't be so hard/dangerous to use these gathering write APIs :-( )

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:289]: I would expect that 0 can only be returned if the channel is configured non blocking. So maybe for our piece of mind we could throw an IOException if that happens, which would then be caught below and cause close() and subscription.cancel() to be called, and the result to be completed exceptionally?

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 2]
```diff
@@ -281,8 +281,13 @@ public void onSubscribe(Flow.Subscription subscription) {
 
         @Override
         public void onNext(List<ByteBuffer> items) {
+            int size = Utils.remaining(items, Integer.MAX_VALUE);
```
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:284]: Hello Chris, are we intentionally limiting the `max` to `Integer.MAX_VALUE` here? This specific `Utils.remaining` will throw an `IllegalArgumentException` if the total `remaining` content across the `items` exceed that max value. If it is intentional to use this method, should we explicitly handle the `IAE` and rethrow it as `IOException`?

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 1 above)
Comment (jaikiran) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:289]: Should we handle the potential case of long value overflow, after addition here, which could then cause the while loop to misbehave?

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 2 above)
Comment (jaikiran) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:284]: There's an overloaded `Utils.remaining` which doesn't constrain the maximum size. Perhaps we could use that?

Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: 👍 Thanks @jaikiran. Agree. Done.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 3]
```diff
@@ -281,8 +281,17 @@ public void onSubscribe(Flow.Subscription subscription) {
...
         public void onNext(List<ByteBuffer> items) {
+            long expectedSize = Utils.remaining(items);
             try {
-                out.write(items.toArray(Utils.EMPTY_BB_ARRAY));
+                long bytesWritten = 0;
+                ByteBuffer[] buffers = items.toArray(Utils.EMPTY_BB_ARRAY);
+                while (bytesWritten < expectedSize) {
+                    long n = out.write(buffers);
+                    if (n <= 0) {
+                        throw new IOException("zero bytes written");
```
Comment (dweiss) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:291]: "Returns: The number of bytes written, possibly zero"

Sorry for being dumb, but why is the returned zero bytes treated as an error? If I were to write a fuzzifier for that class, it'd be my first thing on the to do list to randomly return zero bytes - if the contract allows it, it should be handled gracefully.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 1 above)
Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: I replaced usages with Math::addExact, and generalised the catch to handle arithmetic exceptions too

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: That's the reason I originally didn't comment on the size limitation to Integer.MAX_VALUE - this is an acceptable limitation and it avoids to have to care about overflows here if size <=  Integer.MAX_VALUE. Integer.MAX_VALUE shoudln't be reached in the context where the publisher is the HttpClient.

If we want to allow longs however, we should really use Math.addExact all the way down - including in Utils::remaining.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 3 above)
Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:291]: Well this has the potential of creating a busy loop since bytesWritten will not increase if zero is returned.

File: src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java [hunk 4]
```diff
@@ -668,7 +668,7 @@ public static boolean hasRemaining(List<ByteBuffer> bufs) {
     public static long remaining(List<ByteBuffer> bufs) {
         long remain = 0;
         for (ByteBuffer buf : bufs) {
-            remain += buf.remaining();
+            remain = Math.addExact(remain, buf.remaining());
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/common/Utils.java:671]: There's another occurrence at line 655

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 3 above)
Comment (dweiss) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:291]: I know that. But write should eventually make progress - even if it spins a few times. What's the point of having that JDK method explicitly state "may return 0" and then disregarding that as something that is an exceptional scenario? I don't see the consistency here... If it can return zero, fine - handle it by retrying. If this code falls into a spin loop and never makes progress - this would be a problem somewhere else (perhaps in the implementation of the file channel), not here.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 1 above)
Comment (djelinski) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: I don't think we can ever overflow long here; ByteBuffer.remaining can not exceed Integer.MAX_VALUE, and items.toArray cannot create an array with more elements than integer.MAX_VALUE.

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: Ah! Good point...

Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: True, however Math.addExact is fine here.

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java (hunk 3 above)
Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:291]: I'm on the fence a little on this point. I really dislike the potential for a busy loop, but also don't want the added complication of retry/backoff in this loop. But really the latter is the right solution. lemme see how it looks.

Comment (dweiss) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:291]: Don't get me wrong - you folks are much deeper into low-level details than I am so perhaps it isn't possible that the OS or the hardware ever returns 0 here. From the public API contract perspective there's nothing preventing it from happening though. It's a bit strange for me to see the internal implementation accepts a stricter requirement here.

Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: I added a trivial retry mechanism see https://github.com/openjdk/jdk/pull/11722/commits/c1355274911537fd25eae873908b84617bbb2e58

**Comment by dfuch** [comment]
@ChrisHegarty 

It looks like we upset something :-) 
The Exception catch is probably too wide now.

```
subscribing
test BodySubscribersTest.nulls(ofFile): failure
java.lang.AssertionError: Expected NullPointerException to be thrown, but AssertionError was thrown
	at org.testng.Assert.expectThrows(Assert.java:1724)
	at BodySubscribersTest.nulls(BodySubscribersTest.java:118)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.TestInvoker.invokeMethod(TestInvoker.java:599)
	at org.testng.internal.TestInvoker.invokeTestMethod(TestInvoker.java:174)
	at org.testng.internal.MethodRunner.runInSequence(MethodRunner.java:46)
	at org.testng.internal.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:822)
	at org.testng.internal.TestInvoker.invokeTestMethods(TestInvoker.java:147)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:764)
	at org.testng.TestRunner.run(TestRunner.java:585)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:384)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:378)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:337)
	at org.testng.SuiteRunner.run(SuiteRunner.java:286)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:53)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:96)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1218)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1140)
	at org.testng.TestNG.runSuites(TestNG.java:1069)
	at org.testng.TestNG.run(TestNG.java:1037)
	at com.sun.javatest.regtest.agent.TestNGRunner.main(TestNGRunner.java:93)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at com.sun.javatest.regtest.agent.MainActionHelper$AgentVMRunnable.run(MainActionHelper.java:312)
	at java.base/java.lang.Thread.run(Thread.java:1623)
Caused by: java.lang.AssertionError: null
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.fail(Assert.java:106)
	at BodySubscribersTest$1.cancel(BodySubscribersTest.java:132)
	at java.net.http/jdk.internal.net.http.ResponseSubscribers$PathSubscriber.onNext(ResponseSubscribers.java:307)
	at java.net.http/jdk.internal.net.http.ResponseSubscribers$PathSubscriber.onNext(ResponseSubscribers.java:172)
	at BodySubscribersTest.lambda$nulls$25(BodySubscribersTest.java:118)
	at org.testng.Assert.expectThrows(Assert.java:1714)
	... 29 more
subscribing
```

File: src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java [hunk 5]
```diff
@@ -279,11 +279,30 @@ public void onSubscribe(Flow.Subscription subscription) {
...
+                        retries++;
+                        if (retries > MAX_WRITE_RETRIES) {
+                            throw new IOException("zero bytes written");
+                        }
+                    } else {
+                        retries = 0;
+                        bytesWritten = Math.addExact(bytesWritten, n);
+                    }
+                }
+            } catch (Exception ex) {
```
Comment (dfuch) [inline src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:305]: I believe we should change that line to `catch (IOException | ArithmeicException) {`

Comment (ChrisHegarty) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java:305]: Agreed. Testing it locally before pushing.

Comment (dfuch) [reply src/java.net.http/share/classes/jdk/internal/net/http/ResponseSubscribers.java]: Alternatively it seems that moving the try down to line 292 works too.

**Review by dfuch — APPROVED** [review]
Tests came back green. LGTM!

**Comment by ChrisHegarty** [comment]
> ...
> Why can't it just spin-wait and let the IO subsystem handle buffer writes, whenever they're possible? I'd find an exception from that method surprising - even if the filesystem returns zero written bytes, for whatever reason, I don't think it's an exceptional situation - it follows the contract of that method and is a valid output to me; handle it like any other bytesWritten > 0?

It is acceptable for the gathering write to return zero bytes, and in some contexts it is acceptable to just unconditionally retry over and over, but the HTTP Client implementation is generally more pedantic than that, preferring to err on the side of caution. What is currently in this PR is a reasonable compromise, at the expense of a little more complexity in the code.

**Comment by dweiss** [comment]
Thanks for the explanation, Chris. I remain somewhat unconvinced about the special treatment for http client code here but perhaps I'm overly sensitive... If you care to read on then my concern is that at the API surface the programmer sees "write the response to a path" functionality - plain and simple. A disk-full or an I/O exception due to a hardware problem is easy to understand but an IOException("zero bytes written") showing up in the logs somewhere would probably leave me severely confused as to what actually happened. To me, the API the http client code is actually using to write the response bytes is an implementation detail. If it's using a low-level API that can return odd zero written bytes, it should be able to cope with that and not just panic and throw exceptions to user-space.

So, to summarize, I just hope this zero-written-bytes exception scenario is not something I'll ever see in reality. :)

**Comment by djelinski** [comment]
StackOverflow is telling me that FileChannel writes are always blocking, so a `write` call may only return zero if there are no bytes to write in any of the buffers. This could happen if the same buffer appeared twice on the list, or if some other thread modified the buffers while we were writing. These situations are not very likely to happen.

That being said, I think I'd implement the method as something like:
```
  do {
     out.write(buffers);
  } while (Utils.hasRemaining(buffers));
```
this addresses both the possible failure modes, and leaves the math out of the picture.

**Comment by ChrisHegarty** [comment]
> StackOverflow is telling me that FileChannel writes are always blocking, so a `write` call may only return zero if there are no bytes to write in any of the buffers. This could happen if the same buffer appeared twice on the list, or if some other thread modified the buffers while we were writing. These situations are not very likely to happen.
> 
> That being said, I think I'd implement the method as something like:
> 
> ```
>   do {
>      out.write(buffers);
>   } while (Utils.hasRemaining(buffers));
> ```
> 
> this addresses both the possible failure modes, and leaves the math out of the picture.

This addresses my primary concerns relating to the potential for a busy/infinite loop here - since it short-circuits a complete  iteration of the buffers when determining the loop condition, while still ensuring that all remaining data is written (and avoiding any potential buffer "fiddling" messing up the math).  Darn! I forgot about hasRemaining, nice!

Thanks for this concrete suggestion, accepted. I'm going to add, @djelinski you as a contributor.

@djelinski Would you mind running this latest branch through Mach5, just to ensure that everything is still ok.

/contributor add djelinski

**Review by djelinski — APPROVED** [review]
LGTM. Also, mach5 tier1-3 passed.

**Review by jaikiran — APPROVED** [review]
Hello Chris, these changes look good to me. Please update the copyright years on `Utils.java` and the test class file before integrating.

**Comment by dweiss** [comment]
FYI. Out of curiosity I've looked at how write(ByteBuffer) is implemented for various systems (in native methods in FileDispatcherImpl.c) and I don't see the possibility of zero being returned for a valid write. It's either progress or an error of some sort (EAGAIN in linux, for example).
